cura/Machines/ContainerTree.py

```diff
@@ -76,54 +74,82 @@ def getCurrentQualityChangesGroups(self) -> List["QualityChangesGroup"]:
+        JobQueue.getInstance().add(self.MachineNodeLoadJob(self, currently_added))
+
+    ##  Dictionary-like object that contains the machines.
+    #
+    #   This handles the lazy loading of MachineNodes.
+    class MachineNodeMap:
+        def __init__(self):
+            self._machines = {}
+
+        ##  Returns whether a printer with a certain definition ID exists. This
+        #   is regardless of whether or not the printer is loaded yet.
+        #   \param definition_id The definition to look for.
+        #   \return Whether or not a printer definition exists with that name.
+        def __contains__(self, definition_id: str) -> bool:
+            return len(ContainerRegistry.getInstance().findContainersMetadata(id = definition_id)) > 0
+
+        ##  Returns a machine node for the specified definition ID.
+        #
+        #   If the machine node wasn't loaded yet, this will load it lazily.
+        #   \param definition_id The definition to look for.
+        #   \return A machine node for that definition.
+        def __getitem__(self, definition_id: str) -> MachineNode:
+            if definition_id not in self._machines:
+                start_time = time.time()
+                self._machines[definition_id] = MachineNode(definition_id)
+                self._machines[definition_id].materialsChanged.connect(ContainerTree.getInstance().materialsChanged)
+                Logger.log("d", "Adding container tree for {definition_id} took {duration} seconds.".format(definition_id = definition_id, duration = time.time() - start_time))
+            return self._machines[definition_id]
+
+        ##  Gets a machine node for the specified definition ID, with default.
+        #
+        #   The default is returned if there is no definition with the specified
+        #   ID. If the machine node wasn't loaded yet, this will load it lazily.
+        #   \param definition_id The definition to look for.
+        #   \param default The machine node to return if there is no machine
+        #   with that definition (can be ``None`` optionally or if not
+        #   provided).
+        #   \return A machine node for that definition, or the default if there
+        #   is no definition with the provided definition_id.
+        def get(self, definition_id: str, default: Optional[MachineNode] = None) -> Optional[MachineNode]:
+            if definition_id not in self:
+                return default
+            return self[definition_id]
+
+        ##  Returns whether we've already cached this definition's node.
+        #   \param definition_id The definition that we may have cached.
+        #   \return ``True`` if it's cached.
+        def is_loaded(self, definition_id: str) -> bool:
+            return definition_id in self._machines
+
+    ##  Pre-loads all currently added printers as a background task so that
+    #   switching printers in the interface is faster.
+    class MachineNodeLoadJob(Job):
```

Comment:
Same comment as with the MachineNodeMap

Reply:
Done: f1e35907d255dbc2a148df144982f83a3eeae786